python/tchannel/tornado/tchannel.py

```diff
@@ -159,13 +172,57 @@ def handle_stream(self, stream, address):
         })
 
         log.debug(
-            "Successfully completed handshake with %s (%s)",
+            "Successfully completed handshake with %s:%s (%s)",
             conn.remote_host,
+            conn.remote_host_port,
             conn.remote_process_name,
         )
 
-        self.tchannel.peers.get(conn.remote_host).register_incoming(conn)
+        self.tchannel.peers.get(
+            "%s:%s" % (conn.remote_host,
+                       conn.remote_host_port)
+        ).register_incoming(conn)
+
         yield conn.serve(handler=CallableRequestHandler(self._handle))
 
     def _handle(self, context, connection):
         self.tchannel.receive_call(context, connection)
+
+
+class TChannelProxy(object):
+    """TChannel Proxy with additional runtime info
+
+    TChannelProxy contains parent_tracing information which is created by
+    received request.
+
+    TChannelProxy will be used as one parameter for the request handler.
+
+    Example::
+        def handler(request, response, proxy):
+
+    """
+    __slots__ = ('_tchannel', 'parent_tracing')
+
+    def __init__(self, tchannel, parent_tracing=None):
+        self._tchannel = tchannel
+        self.parent_tracing = parent_tracing
+
+    @property
+    def closed(self):
+        return self._tchannel.closed
+
+    @property
+    def hostport(self):
+        return self._tchannel.hostport
+
+    def host(self, handler):
+        return self._tchannel.host(handler)
+
+    def listen(self):
+        return self._tchannel.listen()
+
+    def request(self, hostport=None, service=None, **kwargs):
+        kwargs['parent_tracing'] = self.parent_tracing
```

Comment:
Consider using [`kwargs.setdefault`](https://docs.python.org/2/library/stdtypes.html#dict.setdefault) instead. That way, the caller can specify `parent_tracing` manually.